Add support for cmd/tlm partials #170

jmthomas opened the pull request:
This isn't ready for merge yet (I'd like to support parameters) but check it out and see what you think.

jmthomas commented:
I tried to implement parameters as close as possible to how rails does it: http://guides.rubyonrails.org/layouts_and_rendering.html#using-partials (see section 3.4.4).

ryanatball commented on lib/cosmos/system/target.rb:
```diff
@@ -186,7 +186,8 @@ def add_all_cmd_tlm(dir)
         Dir[File.join(dir, 'cmd_tlm', '*.txt')].each do |filename|
-          cmd_tlm_files << filename
+          # Don't add partial files which begin with '_'
+          cmd_tlm_files << filename unless filename[0] == '_'
```
The partials need to stay in the cmd_tlm_files list so that they will be included in the MD5 calculation.   You just need to modify System#load_packets to not call process_file on them.  Not sure if that should be based on .erb extension.   See higher level comment

ryanatball replied:
Probably should be based on leading underscore

jmthomas replied:
I ended up letting packet_config.rb decide whether to process the file.

ryanatball commented on lib/cosmos/config/config_parser.rb:
```diff
@@ -158,7 +171,10 @@ def parse_file(filename,
+      # Create a temp file where we can write the ERB parsed output
+      file = Tempfile.new("parsed_#{File.basename(filename)}")
+      file.write(ERB.new(File.read(@filename)).result(binding))
+      file.rewind
```
Not sure if we should run ERB on every file.   Perhaps only if it has a .erb extension (same as rails)?

jmthomas replied:
Right now target.rb only grabs *.txt files in the cmd_tlm directory. Not sure why we did that but obviously we'd have to change that as well. My thought was that we'd just parse our existing text files and if there are things to interpret we do so. I ran this against a complex server configuration and it doesn't seem to really impact the load time. Specifying the extension would allow for future template engines but I'm not sure that really applies to us the same way it does for Rails.

ryanatball commented on lib/cosmos/config/config_parser.rb:
```diff
@@ -143,6 +144,18 @@ def error(message, usage = "", url = @url)
+      b = binding
+      if options[:locals]
+        options[:locals].each do |key, value|
+          eval("@#{key} = #{value}", b)
```
In rails, locals don't get an @ symbol applied to them.  Don't think that is necessary

jmthomas replied:
Yep, changed.

ryanatball commented:
This is an awesome feature!  Well done putting it together so fast

jmthomas commented:
I added a warning about deprecating MACRO_APPEND_START/STOP because it's totally not needed now that we have ERB template support.

ryanatball commented:
I fixed an issue where if you where manually specifying the COMMANDS and TELEMETRY in the target.txt files that partials weren't getting added to the file list.  I also made it support subfolders.   Finally, I removed the parsed_ debugging files as they were being written into areas that I don't want COSMOS touching and I don't want the config_parser.rb file to be dependent on System to look up the tmp path, etc.   

Approved.   It is also just hit me that this is now available in ALL config files which is huge.

jmthomas commented:
All you did was delete the code that writes out the debug files. They were being written right back into the same dir that they originated from. Now if you have an error in your partial you can't see the final result and it's impossible to debug.

ryanatball commented:
It's not impossible, you still get the exception.  There are lots of problems with writing into the target folders. 

1. I consider it to be bad form to write into the target folders as they could be read-only
2. You now silently have two config files that will get parsed next time you startup

If we keep the functionality I would like it to write to the tmp folder.   I guess we could make it try to write to the default tmp folder if it exists, but I don't want config_parser.rb to become dependent on System.   Let me update it to do that.

ryanatball commented:
Also, the parsed_ file may not have any ERB and may be an exact copy of the given config file.   These files get created even for normal configuration errors.

jmthomas commented:
Many good points. Perhaps we just could popup a dialog with the parsed output? That's kind of messy, especially for a long config file. I guess I don't care where the parsed output goes as long as the user can get to it for debugging purposes.

ryanatball commented:
Added an updated working error handler back.  Check it out.

jmthomas commented:
I want that output so I made it always output using the root USERPATH if necessary. Also I added the output filename to the exception message so people know where to go look. Also updated a few config files to use ERB. Why didn't we do this sooner?!?

ryanatball commented:
Did a little more cleanup.   All specs pass.  This is approved.
